client/src/selectionRange.proposed.ts

```diff
@@ -84,84 +84,89 @@ export class SelectionRangeFeature extends TextDocumentFeature<TextDocumentRegis
 		});
 	}
 
-	private asSelectionRanges(selectionRanges: SelectionRange[][] | null): VSelectionRange[][] {
+	private asSelectionRanges(selectionRanges: SelectionRange[] | null): VSelectionRange[] {
 		if (!Array.isArray(selectionRanges)) {
 			return [];
 		}
-		let result:  VSelectionRange[][] = [];
-		for (let ranges of selectionRanges) {
-			if (!Array.isArray(ranges)) {
+		let result: VSelectionRange[] = [];
+		for (let range of selectionRanges) {
+			let vrange = this.asSelectionRange(range);
+			if (vrange == null) {
 				return [];
 			}
-			let inner: VSelectionRange[] = [];
-			result.push(inner);
-			for (let range of ranges) {
-				inner.push(new VSelectionRange(
-					this._client.protocol2CodeConverter.asRange(range.range),
-					this.asSelectionRangeKind(range.kind),
-				));
-			}
+			result.push(vrange);
 		}
 		return result;
 	}
 
-	private asSelectionRangeKind(kind?: string): VSelectionRangeKind {
-		switch (kind) {
-			case SelectionRangeKind.Empty:
-				return VSelectionRangeKind.Empty;
-			case SelectionRangeKind.Statement:
-				return VSelectionRangeKind.Statement;
-			case SelectionRangeKind.Declaration:
-				return VSelectionRangeKind.Declaration;
-			default:
-				return VSelectionRangeKind.Empty
+	private asSelectionRange(selectionRange: SelectionRange): VSelectionRange | undefined {
+		let range = this._client.protocol2CodeConverter.asRange(selectionRange.range);
+		let parent = undefined;
+		if (selectionRange.parent) {
+			parent = this.asSelectionRange(selectionRange.parent);
+			if (!parent || !(parent.range.contains(range) && !parent.range.isEqual(range))) {
+				return undefined
+			}
 		}
+		return new VSelectionRange(range, parent);
 	}
 }
 
 declare module 'vscode' {
```

Comment:
note that the API is not proposed in insiders, so we should remove this once we update vscode dependency